test/cpp/end2end/xds/xds_override_host_end2end_test.cc

```diff
@@ -0,0 +1,241 @@
+      EXPECT_EQ(0, backends_[i]->backend_service1()->request_count())
+          << "Backend " << i << "\n"
+          << debug_location.file() << ":" << debug_location.line();
+      EXPECT_EQ(0, backends_[i]->backend_service2()->request_count())
+          << "Backend " << i << "\n"
+          << debug_location.file() << ":" << debug_location.line();
+    }
+  }
+};
+
+INSTANTIATE_TEST_SUITE_P(XdsTest, OverrideHostTest,
+                         ::testing::Values(XdsTestType()), &XdsTestType::Name);
+TEST_P(OverrideHostTest, HappyPath) {
+  CreateAndStartBackends(2);
+  auto cluster = default_cluster_;
+  cluster.set_lb_policy(Cluster::ROUND_ROBIN);
+  balancer_->ads_service()->SetCdsResource(cluster);
+  SetListenerAndRouteConfiguration(balancer_.get(),
+                                   BuildListenerWithStatefulSessionFilter(),
+                                   default_route_config_);
+  EdsResourceArgs args({{"locality0", CreateEndpointsForBackends()}});
+  balancer_->ads_service()->SetEdsResource(BuildEdsResource(args));
+  balancer_->ads_service()->SetRdsResource(RouteConfiguration());
+  WaitForBackend(DEBUG_LOCATION, 0, /*check_status=*/nullptr,
+                 WaitForBackendOptions());
+  WaitForBackend(DEBUG_LOCATION, 1, /*check_status=*/nullptr,
+                 WaitForBackendOptions());
+  std::vector<std::map<std::string, std::string, std::less<>>> initial_metadata;
+  // First call gets the cookie. RR policy picks the backend we will use.
+  CheckRpcSendOk(DEBUG_LOCATION, 1, RpcOptions(), &initial_metadata);
+  absl::optional<std::string> stateful_session_cookie =
+      GetStatefulSessionCookie(initial_metadata[0]);
+  ASSERT_TRUE(stateful_session_cookie.has_value());
+  size_t backend_idx = -1;
+  for (size_t i = 0; i < backends_.size(); i++) {
+    if (backends_[i]->backend_service()->request_count() == 1) {
+      backends_[i]->backend_service()->ResetCounters();
+      backend_idx = i;
+      break;
+    }
+  }
+  ASSERT_NE(-1, backend_idx);
+  CheckBackendCallCounts(0);
+  // All requests go to the backend we specified
+  CheckRpcSendOk(DEBUG_LOCATION, 5,
+                 RpcOptions().set_metadata(
+                     {{"cookie", absl::StrFormat("%s=%s", "grpc_session_cookie",
+                                                 *stateful_session_cookie)}}),
+                 &initial_metadata);
+  CheckBackendCallCounts(
+      [=](size_t idx) { return idx == backend_idx ? 5 : 0; });
+  // Round-robin spreads the load
+  CheckRpcSendOk(DEBUG_LOCATION, backends_.size() * 2);
+  CheckBackendCallCounts(2);
+  // Call a different service with the same cookie
+  CheckRpcSendOk(
+      DEBUG_LOCATION, 5,
+      RpcOptions()
+          .set_metadata(
+              {{"cookie", absl::StrFormat("%s=%s", "grpc_session_cookie",
+                                          *stateful_session_cookie)}})
+          .set_rpc_service(RpcService::SERVICE_ECHO2),
+      &initial_metadata);
```

Comment:
There's no need to capture the server initial metadata here, since we're not using it for anything.